lib/Translation/ExportVerilog/ExportVerilog.cpp

```diff
@@ -166,6 +168,99 @@ static bool isZeroBitType(Type type) {
   return false;
 }
 
+/// Given a set of known nested types (those supported by this pass), strip off
+/// leading unpacked types.  This strips off portions of the type that are
+/// printed to the right of the name in verilog.
+static Type stripUnpackedTypes(Type type) {
+  return TypeSwitch<Type, Type>(type)
+      .Case<InOutType>([](InOutType inoutType) {
+        return stripUnpackedTypes(inoutType.getElementType());
+      })
+      .Case<UnpackedArrayType>([](UnpackedArrayType arrayType) {
+        return stripUnpackedTypes(arrayType.getElementType());
+      })
+      .Default([](Type type) { return type; });
+}
+
+/// Output the basic type that consists of packed and primitive types.  This is
+/// those to the left of the name in verilog. implicitIntType controls whether
+/// to print a base type for (logic) for inteters or whether the caller will
+/// have handled this (with logic, wire, reg, etc).  structFieldSep is a
+/// character to be printed between struct fields.  Pretty printed structs will
+/// likely use newline, while inline structs will use spaces
+static void printPackedTypeImpl(Type type, raw_ostream &os, Location loc,
+                                SmallVectorImpl<size_t> &dims,
+                                bool implicitIntType, char structFieldSep) {
+  return TypeSwitch<Type, void>(type)
+      .Case<IntegerType>([&](IntegerType integerType) {
+        if (!implicitIntType)
+          os << "logic";
+        if (integerType.getWidth() != 1)
+          dims.push_back(integerType.getWidth());
+        if (!dims.empty() && !implicitIntType)
+          os << ' ';
+
+        for (auto dim : dims)
+          if (dim)
+            os << '[' << (dim - 1) << ":0]";
+          else
+            os << "/*Zero Width*/";
+      })
+      .Case<InOutType>([&](InOutType inoutType) {
+        printPackedTypeImpl(inoutType.getElementType(), os, loc, dims,
+                            implicitIntType, structFieldSep);
+      })
+      .Case<StructType>([&](StructType structType) {
+        os << "struct packed {";
+        for (auto &element : structType.getElements()) {
+          SmallVector<size_t, 8> structDims;
```

Comment:
needless microoptimization: I'd recommend pull the smallvector out of the loop and clear it on each iteration instead of reinitializing it.